tools/terraform/fleetdm_client/fleetdm_client.go

```diff
@@ -0,0 +1,276 @@
+}
+
+type TeamQueryResponse struct {
+	Teams []Team `json:"teams"`
+}
+
+// FleetDMClient is a FleetDM HTTP client that overrides http.DefaultClient.
+type FleetDMClient struct {
+	*http.Client
+	URL    string
+	APIKey string
+}
+
+// NewFleetDMClient creates a new instance of FleetDMClient with the provided
+// URL and API key.
+func NewFleetDMClient(url, apiKey string) *FleetDMClient {
+	return &FleetDMClient{
+		Client: http.DefaultClient,
+		URL:    url,
+		APIKey: apiKey,
+	}
+}
+
+// Do will add necessary headers and call the http.Client.Do method.
+func (c *FleetDMClient) do(req *http.Request, query string) (*http.Response, error) {
+	// Add the API key to the request header
+	req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", c.APIKey))
+	req.Header.Add("Accept", `application/json`)
+	// Set the request URL based on the client URL
+	req.URL, _ = url.Parse(c.URL + req.URL.Path)
+	if query != "" {
+		req.URL.RawQuery = query
+	}
+	// Send the request using the embedded http.Client
+	return c.Client.Do(req)
+}
+
+// TeamNameToId will return the ID of a team given the name.
+func (c *FleetDMClient) TeamNameToId(name string) (int64, error) {
+	req, err := http.NewRequest(http.MethodGet, teamPrefix, nil)
+	if err != nil {
+		return 0, fmt.Errorf("failed to create GET request for %s: %w", teamPrefix, err)
+	}
+	query := fmt.Sprintf("query=%s", name)
+	resp, err := c.do(req, query)
+	if err != nil {
+		return 0, fmt.Errorf("failed to GET %s %s: %w", teamPrefix, query, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return 0, fmt.Errorf("failed to get team: %s %s", query, resp.Status)
+	}
+
+	var teamqry TeamQueryResponse
+	err = json.NewDecoder(resp.Body).Decode(&teamqry)
+	if err != nil {
+		return 0, fmt.Errorf("failed to decode get team response: %w", err)
+	}
+
+	for _, team := range teamqry.Teams {
+		if team.Name == name {
+			return team.ID, nil
+		}
+	}
+
+	return 0, fmt.Errorf("team %s not found", name)
+}
+
+// CreateTeam creates a new team with the provided name and description.
+func (c *FleetDMClient) CreateTeam(name string, description string) (*TeamGetResponse, error) {
+	teamCreate := TeamCreate{
+		Name:        name,
+		Description: description,
+	}
+	nameJson, err := json.Marshal(teamCreate)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create team body: %w", err)
+	}
+	req, err := http.NewRequest(http.MethodPost, teamPrefix, bytes.NewReader(nameJson))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create POST request for %s name %s: %w",
+			teamPrefix, name, err)
+	}
+	resp, err := c.do(req, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to POST %s name %s: %w",
+			teamPrefix, name, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 300 {
+		b, _ := io.ReadAll(resp.Body)
+		return nil, fmt.Errorf("failed to create team %s: %s %s", name, resp.Status, string(b))
+	}
+
+	var newTeam TeamGetResponse
+	err = json.NewDecoder(resp.Body).Decode(&newTeam)
```

Comment:
Same comment here (probably in all client methods) regarding `Decoder` vs `Unmarshal`.

Reply:
Changed (and all below)